File: LogisimPlus/LogisimPlus/logisim/circuit/Wire.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;
using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.circuit
{

	using Component = logisim.comp.Component;
	using ComponentFactory = logisim.comp.ComponentFactory;
	using ComponentDrawContext = logisim.comp.ComponentDrawContext;
	using ComponentListener = logisim.comp.ComponentListener;
	using EndData = logisim.comp.EndData;
	using logisim.data;
	using AttributeListener = logisim.data.AttributeListener;
	using AttributeOption = logisim.data.AttributeOption;
	using AttributeSet = logisim.data.AttributeSet;
	using Attributes = logisim.data.Attributes;
	using BitWidth = logisim.data.BitWidth;
	using Bounds = logisim.data.Bounds;
	using Location = logisim.data.Location;
	using CustomHandles = logisim.tools.CustomHandles;
	using Cache = logisim.util.Cache;
	using GraphicsUtil = logisim.util.GraphicsUtil;

	public sealed class Wire : Component, AttributeSet, CustomHandles, IEnumerable<Location>
	{
		/// <summary>
		/// Stroke width when drawing wires. </summary>
		public const int WIDTH = 3;

		public static readonly AttributeOption VALUE_HORZ = new AttributeOption("horz", Strings.getter("wireDirectionHorzOption"));
		public static readonly AttributeOption VALUE_VERT = new AttributeOption("vert", Strings.getter("wireDirectionVertOption"));
		public static readonly Attribute<AttributeOption> dir_attr = Attributes.forOption("direction", Strings.getter("wireDirectionAttr"), new AttributeOption[] {VALUE_HORZ, VALUE_VERT});
		public static readonly Attribute<int> len_attr = Attributes.forInteger("length", Strings.getter("wireLengthAttr"));

// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
// ORIGINAL LINE: private static final java.util.List<logisim.data.Attribute<?>> ATTRIBUTES = java.util.Arrays.asList(new logisim.data.Attribute<?>[] { dir_attr, len_attr });
		private static readonly IList<Attribute<object>> ATTRIBUTES = new List<Attribute<object>> {dir_attr, len_attr};
		private static readonly Cache cache = new Cache();

		public static Wire create(Location e0, Location e1)
		{
			return (Wire) cache.get(new Wire(e0, e1));
		}

		private class EndList : System.Collections.ObjectModel.Collection<EndData>
		{
			private readonly Wire outerInstance;

			public EndList(Wire outerInstance)
			{
				this.outerInstance = outerInstance;
			}

			public override EndData get(int i)
			{
				return outerInstance.getEnd(i);
			}

			public override int size()
			{
				return 2;
			}
		}

		internal readonly Location e0;
		internal readonly Location e1;
		internal readonly bool is_x_equal;

		private Wire(Location e0, Location e1)
		{
			this.is_x_equal = e0.X == e1.X;
			if (is_x_equal)
			{
				if (e0.Y > e1.Y)
				{
					this.e0 = e1;
					this.e1 = e0;
				}
				else
				{
					this.e0 = e0;
					this.e1 = e1;
				}
			}
			else
			{
				if (e0.X > e1.X)
				{
					this.e0 = e1;
					this.e1 = e0;
				}
				else
				{
					this.e0 = e0;
					this.e1 = e1;
				}
			}
		}

		public override bool Equals(object other)
		{
			if (!(other is Wire))
			{
				return false;
			}
			Wire w = (Wire) other;
			return w.e0.Equals(this.e0) && w.e1.Equals(this.e1);
		}

		public override int GetHashCode()
		{
			return e0.GetHashCode() * 31 + e1.GetHashCode();
		}

		public int Length
		{
			get
			{
				return (e1.Y - e0.Y) + (e1.X - e0.X);
			}
		}

		public override string ToString()
		{
			return "Wire[" + e0 + "-" + e1 + "]";
		}

		//
		// Component methods
		//
		// (Wire never issues ComponentEvents, so we don't need to track listeners)
		public void addComponentListener(ComponentListener e)
		{
		}

		public void removeComponentListener(ComponentListener e)
		{
		}

		public ComponentFactory Factory
		{
			get
			{
				return WireFactory.instance;
			}
		}

		public AttributeSet AttributeSet
		{
			get
			{
				return this;
			}
		}

		// location/extent methods
		public Location Location
		{
			get
			{
				return e0;
			}
		}

		public Bounds Bounds
		{
			get
			{
				int x0 = e0.X;
				int y0 = e0.Y;
				return Bounds.create(x0 - 2, y0 - 2, e1.X - x0 + 5, e1.Y - y0 + 5);
			}
		}

		public Bounds getBounds(Graphics g)
		{
			return Bounds;
		}

		public bool contains(Location q)
		{
			int qx = q.X;
			int qy = q.Y;
			if (is_x_equal)
			{
				int wx = e0.X;
				return qx >= wx - 2 && qx <= wx + 2 && e0.Y <= qy && qy <= e1.Y;
			}
			else
			{
				int wy = e0.Y;
				return qy >= wy - 2 && qy <= wy + 2 && e0.X <= qx && qx <= e1.X;
			}
		}

		public bool contains(Location pt, Graphics g)
		{
			return contains(pt);
		}

		//
		// propagation methods
		//
		public IList<EndData> Ends
		{
			get
			{
				return new EndList(this);
			}
		}

		public EndData getEnd(int index)
		{
			Location loc = getEndLocation(index);
			return new EndData(loc, BitWidth.UNKNOWN, EndData.INPUT_OUTPUT);
		}

		public bool endsAt(Location pt)
		{
			return e0.Equals(pt) || e1.Equals(pt);
		}

		public void propagate(CircuitState state)
		{
			// Normally this is handled by CircuitWires, and so it won't get
			// called. The exception is when a wire is added or removed
			state.markPointAsDirty(e0);
			state.markPointAsDirty(e1);
		}

		//
		// user interface methods
		//
		public void expose(ComponentDrawContext context)
		{
			java.awt.Component dest = context.Destination;
			int x0 = e0.X;
			int y0 = e0.Y;
			dest.repaint(x0 - 5, y0 - 5, e1.X - x0 + 10, e1.Y - y0 + 10);
		}

		public void draw(ComponentDrawContext context)
		{
			CircuitState state = context.CircuitState;
			Graphics g = context.Graphics;

			GraphicsUtil.switchToWidth(g, WIDTH);
			g.setColor(state.getValue(e0).Color);
			g.drawLine(e0.X, e0.Y, e1.X, e1.Y);
		}

		public object getFeature(object key)
		{
			if (key == typeof(CustomHandles))
			{
				return this;
			}
			return null;
		}

		//
		// AttributeSet methods
		//
		// It makes some sense for a wire to be its own attribute, since
		// after all it is immutable.
		//
		public object clone()
		{
			return this;
		}

		public void addAttributeListener(AttributeListener l)
		{
		}

		public void removeAttributeListener(AttributeListener l)
		{
		}

// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
// ORIGINAL LINE: public java.util.List<logisim.data.Attribute<?>> getAttributes()
		public IList<Attribute<object>> Attributes
		{
			get
			{
				return ATTRIBUTES;
			}
		}

		public bool containsAttribute<T1>(Attribute<T1> attr)
		{
			return ATTRIBUTES.Contains(attr);
		}

// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
// ORIGINAL LINE: public logisim.data.Attribute<?> getAttribute(String name)
		public Attribute<object> getAttribute(string name)
		{
// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
// ORIGINAL LINE: for (logisim.data.Attribute<?> attr : ATTRIBUTES)
			foreach (Attribute<object> attr in ATTRIBUTES)
			{
				if (name.Equals(attr.Name))
				{
					return attr;
				}
			}
			return null;
		}

		public bool isReadOnly<T1>(Attribute<T1> attr)
		{
			return true;
		}

		public void setReadOnly<T1>(Attribute<T1> attr, bool value)
		{
			throw new System.NotSupportedException();
		}

		public bool isToSave<T1>(Attribute<T1> attr)
		{
			return false;
		}

// JAVA TO C# CONVERTER TASK: Most Java annotations will not have direct .NET equivalent attributes:
// ORIGINAL LINE: @SuppressWarnings("unchecked") public <V> V getValue(logisim.data.Attribute<V> attr)
		public V getValue<V>(Attribute<V> attr)
		{
			if (attr == dir_attr)
			{
				return (V)(is_x_equal ? VALUE_VERT : VALUE_HORZ);
			}
			else if (attr == len_attr)
			{
				return (V) Convert.ToInt32(Length);
			}
			else
			{
				return null;
			}
		}

		public void setValue<V>(Attribute<V> attr, V value)
		{
			throw new System.ArgumentException("read only attribute");
		}

		//
		// other methods
		//
		public bool Vertical
		{
			get
			{
				return is_x_equal;
			}
		}

		public Location getEndLocation(int index)
		{
			return index == 0 ? e0 : e1;
		}

		public Location End0
		{
			get
			{
				return e0;
			}
		}

		public Location End1
		{
			get
			{
				return e1;
			}
		}

		public Location getOtherEnd(Location loc)
		{
			return (loc.Equals(e0) ? e1 : e0);
		}

		public bool sharesEnd(Wire other)
		{
			return this.e0.Equals(other.e0) || this.e1.Equals(other.e0) || this.e0.Equals(other.e1) || this.e1.Equals(other.e1);
		}

		public bool overlaps(Wire other, bool includeEnds)
		{
			return overlaps(other.e0, other.e1, includeEnds);
		}

		private bool overlaps(Location q0, Location q1, bool includeEnds)
		{
			if (is_x_equal)
			{
				int x0 = q0.X;
				if (x0 != q1.X || x0 != e0.X)
				{
					return false;
				}
				if (includeEnds)
				{
					return e1.Y >= q0.Y && e0.Y <= q1.Y;
				}
				else
				{
					return e1.Y > q0.Y && e0.Y < q1.Y;
				}
			}
			else
			{
				int y0 = q0.Y;
				if (y0 != q1.Y || y0 != e0.Y)
				{
					return false;
				}
				if (includeEnds)
				{
					return e1.X >= q0.X && e0.X <= q1.X;
				}
				else
				{
					return e1.X > q0.X && e0.X < q1.X;
				}
			}
		}

		public bool isParallel(Wire other)
		{
			return this.is_x_equal == other.is_x_equal;
		}

		public IEnumerator<Location> GetEnumerator()
		{
			return new WireIterator(e0, e1);
		}

		public void drawHandles(ComponentDrawContext context)
		{
			context.drawHandle(e0);
			context.drawHandle(e1);
		}
	}

}

File: logisim/app/src/main/java/logisim/circuit/Wire.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.circuit;

import java.awt.Graphics;
import java.util.AbstractList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;

import logisim.comp.Component;
import logisim.comp.ComponentFactory;
import logisim.comp.ComponentDrawContext;
import logisim.comp.ComponentListener;
import logisim.comp.EndData;
import logisim.data.Attribute;
import logisim.data.AttributeListener;
import logisim.data.AttributeOption;
import logisim.data.AttributeSet;
import logisim.data.Attributes;
import logisim.data.BitWidth;
import logisim.data.Bounds;
import logisim.data.Location;
import logisim.tools.CustomHandles;
import logisim.util.Cache;
import logisim.util.GraphicsUtil;
import org.jetbrains.annotations.NotNull;

public final class Wire implements Component, AttributeSet, CustomHandles, Iterable<Location> {
	/** Stroke width when drawing wires. */
	public static final int WIDTH = 3;

	public static final AttributeOption VALUE_HORZ = new AttributeOption("horz",
			Strings.getter("wireDirectionHorzOption"));
	public static final AttributeOption VALUE_VERT = new AttributeOption("vert",
			Strings.getter("wireDirectionVertOption"));
	public static final Attribute<AttributeOption> dir_attr = Attributes.forOption("direction",
			Strings.getter("wireDirectionAttr"), new AttributeOption[] { VALUE_HORZ, VALUE_VERT });
	public static final Attribute<Integer> len_attr = Attributes.forInteger("length", Strings.getter("wireLengthAttr"));

	private static final List<Attribute<?>> ATTRIBUTES = Arrays.asList(dir_attr, len_attr);
	private static final Cache cache = new Cache();

	public static Wire create(Location e0, Location e1) {
		return (Wire) cache.get(new Wire(e0, e1));
	}

	private class EndList extends AbstractList<EndData> {
		@Override
		public EndData get(int i) {
			return getEnd(i);
		}

		@Override
		public int size() {
			return 2;
		}
	}

	final Location e0;
	final Location e1;
	final boolean is_x_equal;

	private Wire(Location e0, Location e1) {
		is_x_equal = e0.x() == e1.x();
		if (is_x_equal) if (e0.y() > e1.y()) {
			this.e0 = e1;
			this.e1 = e0;
		}
		else {
			this.e0 = e0;
			this.e1 = e1;
		}
		else if (e0.x() > e1.x()) {
			this.e0 = e1;
			this.e1 = e0;
		}
		else {
			this.e0 = e0;
			this.e1 = e1;
		}
	}

	@Override
	public boolean equals(Object other) {
		return other instanceof Wire w && w.e0.equals(e0) && w.e1.equals(e1);
	}

	@Override
	public int hashCode() {
		return e0.hashCode() * 31 + e1.hashCode();
	}

	public int getLength() {
		return (e1.y() - e0.y()) + (e1.x() - e0.x());
	}

	@Override
	public String toString() {
		return "Wire[" + e0 + "-" + e1 + "]";
	}

	//
	// Component methods
	//
	// (Wire never issues ComponentEvents, so we don't need to track listeners)
	public void addComponentListener(ComponentListener e) {
	}

	public void removeComponentListener(ComponentListener e) {
	}

	public ComponentFactory getFactory() {
		return WireFactory.instance;
	}

	public AttributeSet getAttributeSet() {
		return this;
	}

	// location/extent methods
	public Location getLocation() {
		return e0;
	}

	public Bounds getBounds() {
		int x0 = e0.x();
		int y0 = e0.y();
		return Bounds.create(x0 - 2, y0 - 2, e1.x() - x0 + 5, e1.y() - y0 + 5);
	}

	public Bounds getBounds(Graphics g) {
		return getBounds();
	}

	public boolean contains(Location q) {
		int qx = q.x();
		int qy = q.y();
		if (is_x_equal) {
			int wx = e0.x();
			return qx >= wx - 2 && qx <= wx + 2 && e0.y() <= qy && qy <= e1.y();
		} else {
			int wy = e0.y();
			return qy >= wy - 2 && qy <= wy + 2 && e0.x() <= qx && qx <= e1.x();
		}
	}

	public boolean contains(Location pt, Graphics g) {
		return contains(pt);
	}

	//
	// propagation methods
	//
	public List<EndData> getEnds() {
		return new EndList();
	}

	public EndData getEnd(int index) {
		Location loc = getEndLocation(index);
		return new EndData(loc, BitWidth.UNKNOWN, EndData.INPUT_OUTPUT);
	}

	public boolean endsAt(Location pt) {
		return e0.equals(pt) || e1.equals(pt);
	}

	public void propagate(CircuitState state) {
		// Normally this is handled by CircuitWires, and so it won't get
		// called. The exception is when a wire is added or removed
		state.markPointAsDirty(e0);
		state.markPointAsDirty(e1);
	}

	//
	// user interface methods
	//
	public void expose(ComponentDrawContext context) {
		java.awt.Component dest = context.getDestination();
		int x0 = e0.x();
		int y0 = e0.y();
		dest.repaint(x0 - 5, y0 - 5, e1.x() - x0 + 10, e1.y() - y0 + 10);
	}

	public void draw(ComponentDrawContext context) {
		CircuitState state = context.getCircuitState();
		Graphics g = context.getGraphics();

		GraphicsUtil.switchToWidth(g, WIDTH);
		g.setColor(state.getValue(e0).getColor());
		g.drawLine(e0.x(), e0.y(), e1.x(), e1.y());
	}

	public Object getFeature(Object key) {
		if (key == CustomHandles.class)
			return this;
		return null;
	}

	//
	// AttributeSet methods
	//
	// It makes some sense for a wire to be its own attribute, since
	// after all it is immutable.
	//
	@Override
	public Object clone() {
		return this;
	}

	public void addAttributeListener(AttributeListener l) {
	}

	public void removeAttributeListener(AttributeListener l) {
	}

	public List<Attribute<?>> getAttributes() {
		return ATTRIBUTES;
	}

	public boolean containsAttribute(Attribute<?> attr) {
		return ATTRIBUTES.contains(attr);
	}

	public Attribute<?> getAttribute(String name) {
		for (Attribute<?> attr : ATTRIBUTES)
			if (name.equals(attr.getName()))
				return attr;
		return null;
	}

	public boolean isReadOnly(Attribute<?> attr) {
		return true;
	}

	public void setReadOnly(Attribute<?> attr, boolean value) {
		throw new UnsupportedOperationException();
	}

	public boolean isToSave(Attribute<?> attr) {
		return false;
	}

	@SuppressWarnings("unchecked")
	public <V> V getValue(Attribute<V> attr) {
		if (attr == dir_attr) return (V) (is_x_equal ? VALUE_VERT : VALUE_HORZ);
		else if (attr == len_attr) return (V) Integer.valueOf(getLength());
		else return null;
	}

	public <V> void setValue(Attribute<V> attr, V value) {
		throw new IllegalArgumentException("read only attribute");
	}

	//
	// other methods
	//
	public boolean isVertical() {
		return is_x_equal;
	}

	public Location getEndLocation(int index) {
		return index == 0 ? e0 : e1;
	}

	public Location getEnd0() {
		return e0;
	}

	public Location getEnd1() {
		return e1;
	}

	public Location getOtherEnd(Location loc) {
		return (loc.equals(e0) ? e1 : e0);
	}

	public boolean sharesEnd(Wire other) {
		return e0.equals(other.e0) || e1.equals(other.e0) || e0.equals(other.e1)
				|| e1.equals(other.e1);
	}

	public boolean overlaps(Wire other, boolean includeEnds) {
		return overlaps(other.e0, other.e1, includeEnds);
	}

	private boolean overlaps(Location q0, Location q1, boolean includeEnds) {
		if (is_x_equal) {
			int x0 = q0.x();
			if (x0 != q1.x() || x0 != e0.x())
				return false;
			if (includeEnds) return e1.y() >= q0.y() && e0.y() <= q1.y();
			else return e1.y() > q0.y() && e0.y() < q1.y();
		} else {
			int y0 = q0.y();
			if (y0 != q1.y() || y0 != e0.y())
				return false;
			if (includeEnds) return e1.x() >= q0.x() && e0.x() <= q1.x();
			else return e1.x() > q0.x() && e0.x() < q1.x();
		}
	}

	public boolean isParallel(Wire other) {
		return is_x_equal == other.is_x_equal;
	}

	public @NotNull Iterator<Location> iterator() {
		return new WireIterator(e0, e1);
	}

	public void drawHandles(ComponentDrawContext context) {
		context.drawHandle(e0);
		context.drawHandle(e1);
	}
}
